### ISSUE TITLE
PW 3.0.158: has_parent in selector misses a lot of pages 

### ISSUE THREAD

**Issue by processwired** [processwire/processwire-issues#1185]
### Short description of the issue

I've used a selector like $pages->find('has_parent=123') and was surprised that it found three pages. After [googling around](https://processwire.com/talk/topic/8585-issue-with-has_parent-selector/) I found that this seems a mysterious bug: this time I fell in this trap :)

After trashing the parent (id 123) and restore it from trash, everything worked as expected. I made a diff between the two databases (before/after trashing and restoring) and I can clearly see that the pages_parents table has many more entries after that.

I think that's really a critical bug, because I can't rely on the data anymore - I used has_parent to delete critical pages with user submitted personal data, and now I can't tell what should have been deleted in the past and what is still there.

### Setup/Environment

* ProcessWire version: 3.0.155, 3.0.157

**Comment by adrianbj** [comment]
Are you really talking about 3.0.55 (which is about 3-4 years old) or do you mean 3.0.155 (which is about a month old)?

Note that there were significant changes to the "parents" system in 3.0.156 (https://processwire.com/talk/topic/23664-pw-30156-%E2%80%93-core-updates/) so it would be good to know if this has fixed the problem for you.

**Comment by processwired** [comment]
How could I have missed that "1"? My mistake, you're absolutely right.

**Comment by processwired** [comment]
Ah, and to answer your question: the problem still exists in **3.0.157**.

**Comment by adrianbj** [comment]
Thanks for clarifying - I was hoping maybe Ryan's hard work in redoing the parents system might have fixed it already.

I must admin that the trash system is starting to worry me because it affects the behavior of other selectors that you don't notice until a page is trashed. Theoretically all this is "in the manual", but sometimes it's hard to remember what selectors do and don't include trashed pages, eg the new `has()` method.

**Comment by processwired** [comment]
Yes, I suppose that the issue has something to do with the trash. I think I can reproduce the issue (which does not mean that I have any idea what's going on :)

I deleted just 1 page, which were an indirect child of id 123 (just to keep that example from above).

After that I created a new page (by API, not in admin), which is an indirect children of 123 some levels down the page tree (quite like the one which was trashed before), and again: this page is not found with "has_parent=123".

Then I tried the "move parent to trash/restore“ trick, now the new page is found.

All tests have been done with 3.0.**157** (latest dev release, just installed a couple of minutes ago).

**Comment by ryancramerdesign** [comment]
@processwired  I've been continuing tweaks on the new PagesParents class, and there may still be issues to resolve yet (it's the dev branch and it's all new code), but I came across a similar issue this week and assuming what you found is the same, it should be fixed in 3.0.158 (today's version), but let me know if not. 

Just updating the PW version doesn't change anything in the page_parents table. So it may be good to start with a fresh copy of the pages_parents table, which you can get by doing this: 
~~~~~
$pages->parents()->rebuildAll();
~~~~~

**Comment by processwired** [comment]
Using $pages->parents()->rebuildAll(); did the trick. The pages are found now. I'm doing some more in-depth tests now. If I find anything, I'll report back here.

@ryancramerdesign I think I sound like a broken record player, but I can only repeat myself: your fast & continuous support and your efforts to make ProcessWire as perfect as possible are fantastic. I admire that.

**Comment by ryancramerdesign** [comment]
Thanks @processwired I appreciate it. :)

**Comment by processwired** [comment]
I tried some things with 3.0.158, and there are still problems with the page_parents table. In my case, I trashed some pages via API and found that numDescendants() returns a wrong page count afterwards: the counts are too high. My guess is that the trashed pages are still contained in page_parents table. After calling $pages->parents()->rebuildAll() the counts where correct again.

Since my app (which is in production already) shows some statistics to the users, I had to go back to 3.0.155, which works fine so far.
